opm/autodiff/GraphColoring.hpp

```diff
@@ -0,0 +1,244 @@
+/*
+  Copyright 2018 Equinor
+
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+#ifndef OPM_GRAPHCOLORING_HEADER_INCLUDED
+#define OPM_GRAPHCOLORING_HEADER_INCLUDED
+
+#include <vector>
+#include <deque>
+#include <tuple>
+#include <algorithm>
+#include <numeric>
+#include <queue>
+
+namespace Opm
+{
+
+namespace Detail
+{
+template<class Graph>
+std::size_t colorGraphWelshPowell(const Graph& graph,
+                           std::deque<typename Graph::VertexDescriptor>& orderedVertices,
+                           std::vector<int>& colors,
+                           int color, int noVertices)
+{
+    std::vector<int> forbidden(noVertices, false);
+    std::size_t noColored = 0;
+
+    for(auto vertex = orderedVertices.begin(),
+            vertexEnd = orderedVertices.end();
+        vertex != vertexEnd; ++vertex)
+    {
+        // Skip forbidden vertices
+        while(vertex != vertexEnd && forbidden[*vertex])
+            ++vertex;
+        if ( vertex == vertexEnd )
+        {
+            break;
+        }
+
+        // Color Vertex
+        colors[*vertex] = color;
+        ++noColored;
+        // Forbid neighors
+        for(auto edge = graph.beginEdges(*vertex), endEdge = graph.endEdges(*vertex);
+            edge != endEdge; ++edge)
+        {
+            forbidden[edge.target()] = true;
+        }
+    }
+    // forbidden vertices will be colored next for coloring
+    using Vertex = typename Graph::VertexDescriptor;
+    auto newEnd = std::remove_if(orderedVertices.begin(), orderedVertices.end(),
+                                 [&forbidden](const Vertex& vertex)
+                                 {
+                                     return !forbidden[vertex];
+                                 });
+    orderedVertices.resize(newEnd-orderedVertices.begin());
+    return noColored;
+}
+template<class Graph, class Functor>
+std::size_t breadthFirstSearch(const Graph& graph, typename Graph::VertexDescriptor root,
+                        Functor functor)
+{
+    std::vector<int> visited(graph.maxVertex() + 1, false);
```

Comment:
I don't fully understand the reason for using a `vector<int>` to hold `true`/`false` values.  Would you care to elaborate?

Reply:
Speed and safety. At least in the past vector<bool> was kind of special as it was optimized for memory consumption and did not meet all requirements of an STL container. Of course I could have used vector<char> instead.

Reply:
> Speed and safety.

Right.  Thanks for the information.

> At least in the past `vector<bool>` was kind of special

It still is.  The standardisation committee nowadays recognises the special semantic requirements of `vector<bool>` as a mistake.

> Of course I could have used `vector<char>` instead.

We'll consider that as an opportunity for future optimisation if this turns out to be a bottleneck.  For now I'm happy with reasoning.